### PR TITLE
Strip down Talon

### PR DESCRIPTION
**Pull request by umairwaheed**
Allow Talon to be installed without ML.

### CODE REVIEW

**Comment by umairwaheed** [comment]
This is a work in progress at the moment.

**Comment by umairwaheed** [comment]
Ready for review. This PR also makes the non-ML part Python 3 compatible.

**Comment by obukhov-sergey** [comment]
@mailgun-ci test this please

File: talon/quotations.py [hunk 1]
```diff
@@ -13,6 +14,8 @@
 
 from talon.utils import get_delimiter, html_to_text
 from talon import html_quotations
+from six.moves import range
+import six
```
Comment (obukhov-sergey) [inline talon/quotations.py]: Can you add it to setup.py please?

File: talon/signature/learning/featurespace.py [hunk 2]
```diff
@@ -7,9 +7,12 @@
 applying features to them.
 """
 
+from __future__ import absolute_import
 from talon.signature.constants import (SIGNATURE_MAX_LINES,
                                        TOO_LONG_SIGNATURE_LINE)
 from talon.signature.learning.helpers import *
+from six.moves import zip
```
Comment (obukhov-sergey) [inline talon/signature/learning/featurespace.py]: Is it used anywhere?

File: talon/signature/learning/featurespace.py [hunk 3]
```diff
@@ -7,9 +7,12 @@
...
+from functools import reduce
```
Comment (obukhov-sergey) [inline talon/signature/learning/featurespace.py]: Can you please remove it if it's not used anywhere

File: talon/utils.py [hunk 4]
```diff
@@ -46,8 +48,8 @@ def to_unicode(str_or_unicode, precise=False):
     If `precise` flag is True, tries to guess the correct encoding first.
     """
     encoding = quick_detect_encoding(str_or_unicode) if precise else 'utf-8'
```
Comment (obukhov-sergey) [inline talon/utils.py]: Can you please move it inside the ``if`` statement below: with the ``assert`` added to ``quick_detect_encoding`` and ``detect_encoding`` this function raises the ``AsserionException`` if the passed string is already in unicode.

**Comment by obukhov-sergey** [comment]
@umairwaheed thank you very much for the PR! Looks good to me except for the comments.

File: talon/quotations.py (hunk 1 above)
Comment (umairwaheed) [inline talon/quotations.py]: Sure.

File: talon/signature/learning/featurespace.py (hunk 2 above)
Comment (umairwaheed) [inline talon/signature/learning/featurespace.py]: Yep, it is used [here](https://github.com/umairwaheed/talon/blob/cefbcffd599bc10e6a7cfc692f33081861919474/talon/signature/learning/featurespace.py#L77). We are just replacing the `zip` function in the standard library of Python with the `zip` function in `six`.

File: talon/signature/learning/featurespace.py (hunk 3 above)
Comment (umairwaheed) [inline talon/signature/learning/featurespace.py]: Used [here](https://github.com/umairwaheed/talon/blob/cefbcffd599bc10e6a7cfc692f33081861919474/talon/signature/learning/featurespace.py#L77). These functions are replaced with the version found in `six`.

File: talon/utils.py (hunk 4 above)
Comment (umairwaheed) [inline talon/utils.py]: Done.

**Comment by umairwaheed** [comment]
@obukhov-sergey, no problem and thanks for looking into this. Updated the PR acceding to the feedback. Please review.

File: talon/signature/learning/featurespace.py (hunk 3 above)
Comment (obukhov-sergey) [inline talon/signature/learning/featurespace.py]: Got it, thanks

File: talon/signature/learning/featurespace.py (hunk 2 above)
Comment (obukhov-sergey) [inline talon/signature/learning/featurespace.py]: ok, thanks for explanation - I suspected something like this was happening :)

**Comment by obukhov-sergey** [comment]
@mailgun-ci test this please

**Comment by wasabigeek** [comment]
@umairwaheed how do I use the `no-ml` option?

**Comment by umairwaheed** [comment]
@wasabigeek, you can do `python setup.py install --no-ml`.

**Comment by wasabigeek** [comment]
thanks @umairwaheed - it doesn't work with pip I suppose

**Comment by umairwaheed** [comment]
@wasabigeek, with pip you can do `pip install talon --install-option="--no-ml"`. You should add `--no-deps` option to avoid installing requirements automatically.

**Pull request merged**